app/models/blood_pressure_rollup.rb

```diff
@@ -0,0 +1,74 @@
+class BloodPressureRollup < ApplicationRecord
+  upsert_keys ["patient_id", "period_number", "period_type", "year"]
+
+  belongs_to :blood_pressure
+  belongs_to :patient
+  belongs_to :assigned_facility, class_name: "Facility"
+  belongs_to :blood_pressure_facility, class_name: "Facility"
+
+  enum period_type: {month: 0, quarter: 1}
+
+  validates :diastolic, presence: true
+  validates :systolic, presence: true
+  validates :period_number, presence: true, numericality: {only_integer: true}
+  validates :period_type, presence: true
+  validates :year, presence: true, numericality: {greater_than_or_equal_to: 2000, only_integer: true}
+
+  def self.controlled_in_month(time)
+    range = [time.month, time.advance(months: -1).month, time.advance(months: -2).month]
+    sql = <<-SQL
+      SELECT count(1)
+      FROM (
+        SELECT DISTINCT ON (patient_id) *
+        FROM #{table_name}
+        WHERE period_number in (?)
+        AND period_type = 0
+        ORDER BY patient_id ASC, (year, period_number) DESC
+      ) AS counts
+      WHERE diastolic >= 90
+      and systolic >= 140
+    SQL
+    query = sanitize_sql_array([sql, range])
+    connection.select_all(query).to_hash
+  end
+
+  def self.from_blood_pressure(blood_pressure)
+    attrs = {
+      assigned_facility_id: blood_pressure.patient.registration_facility_id,
+      blood_pressure_facility_id: blood_pressure.facility_id,
+      blood_pressure_id: blood_pressure.id,
+      diastolic: blood_pressure.diastolic,
+      patient_id: blood_pressure.patient_id,
+      recorded_at: blood_pressure.recorded_at,
+      systolic: blood_pressure.systolic,
+      year: blood_pressure.recorded_at.year
+    }
+
+    month = blood_pressure.recorded_at.month
+    quarter = quarter_for_month(month)
+
+    month_attrs = attrs.merge(period_type: :month, period_number: month)
+    quarter_attrs = attrs.merge(period_type: :quarter, period_number: quarter)
```

Comment:
Think we could split this up into `base_attrs`, `month_attrs`, and `quarter_attrs`?

```ruby
month_rollup=BloodPressureRollup.new(**base_attrs, **month_attrs)
quarter_rollup=BloodPressureRollup.new(**base_attrs, **quarter_attrs)

def base_attrs
  # stuff
end

def month_attrs
  # stuff
end

def quarter_attrs
  # stuff
end
```

Reply:
yeah I plan on refactoring all of this at some point, its pretty ugly right now.